[Feature] Support structural tag templates.

Seven-Streams opened the pull request:
This PR supports the structural tag templates, provides a method to generate a grammar from a structural tag template and the given values, and provides built-in structural tag templates for several models' tool-calling formats.

Ubospica commented on python/xgrammar/structural_tag.py:
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
+    return decorator
+
+
+def get_builtin_template_structural_tag(format_type: str) -> Dict[str, Any]:
```
```suggestion
def get_builtin_template_structural_tag(format_name: str) -> StructuralTag:
```

Ubospica commented on python/xgrammar/structural_tag.py:
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
+
+
+@_register_template_structural_tag_format("Llama")
+def get_llama_style_template_structural_tag() -> Dict[str, Any]:
```
```suggestion
def _get_llama_style_template_structural_tag() -> Dict[str, Any]:
```

Ubospica commented on python/xgrammar/structural_tag.py:
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
+    return func()
+
+
+@_register_template_structural_tag_format("Llama")
```
```suggestion
@_register_template_structural_tag_format("llama")
```

Ubospica commented on python/xgrammar/structural_tag.py:
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
+    }
+
+
+@_register_template_structural_tag_format("Qwen_Coder")
```
```suggestion
@_register_template_structural_tag_format("qwen_coder")
```

Ubospica commented on docs/tutorials/template_structural_tag.md:
```diff
@@ -0,0 +1,133 @@
+
+Based on the [structural tags](./structural_tag.md), template structural tags provides a more convineient method for users to generate grammars to describe the grammar to constrain the LLMs' output.
+In general, Template structural tags support placeholders in the structural tags, which are in the form of `{lists[].arg}`. These values will be automatically expanded with the user's input values. Users can use `xgrammar.Grammar.apply_structural_tag_template(template_json_str: Union[str, Dict[str, Any]], **kwargs: List[Dict[str, Any]])` to generate a `xgrammar.Grammar`. Here, `template_json_str` ...
+
```
Provide a code snippet for usage
1. Tool
2. Builtin Tool

Ubospica commented on python/xgrammar/structural_tag.py:
```diff
@@ -340,3 +340,259 @@ def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
+        A template structural tag format dictionary.
+        This format is used by Qwen3 and other models that follow the same style.
+    """
+    return {
```
add reasoning format to it

1. qwen
2. qwen_non_thinking

Ubospica commented on python/xgrammar/grammar.py:
```diff
@@ -431,3 +432,65 @@ def deserialize_json(json_string: str) -> "Grammar":
         return Grammar._create_from_handle(_core.Grammar.deserialize_json(json_string))
+
+    @staticmethod
+    def apply_structural_tag_template(
```
Grammar.from_structural_tag_template(...)

Ubospica commented on python/xgrammar/builtin_structural_tag_template.py:
```diff
@@ -0,0 +1,246 @@
+_structural_tag_registry = {}
+
+
+def _register_template_structural_tag_format(name: str):
```
```suggestion
def _register_structural_tag_template(name: str):
```

Ubospica commented on python/xgrammar/structural_tag.py:
```diff
@@ -61,7 +61,7 @@ class QwenXMLParameterFormat(BaseModel):
-    json_schema: Union[bool, Dict[str, Any]]
+    json_schema: Union[bool, Dict[str, Any], str]
     """The JSON schema for the parameters of the function calling."""
 
```
StructuralTag.from_template(template, **kwargs)

Ubospica commented on docs/tutorials/structural_tag_template.md:
````diff
@@ -0,0 +1,226 @@
+    "value": "{{A[].begin}} a dog{{A[].end}}"
+}
+```
+
````
Python code: one line, 
```python
structural_tag = StructuralTag.from_template(..., A=xxx)
```

Converted Structural Tag:
```json
{
  "type": "or",
   ...
}
```

Ubospica reviewed (commented):
Thanks! Generally looks good. Let's measure conversion time & compile time for 5 tools & 20 tools

Ubospica reviewed (commented):
Thanks. Let's update tests and we can merge

Ubospica commented on docs/tutorials/structural_tag_template.md:
```diff
@@ -0,0 +1,382 @@
+# Structural Tag Templates
+
+Based on the [structural tags](./structural_tag.md), structural tag templates provides a more convineient method for users to generate grammars to describe the grammar to constrain the LLMs' output.
+In general, structural tag templates support placeholders in the structural tags, which are in the form of `\{\{(var_name([])?)(\.var_name([])?)\}\}`, such as `{{A.name}}`, `{{A[].name}}`, `{{A.name[].other}}`. `var_name[]` means that the value is an array, and it will be automatically expanded with the users' input. `var_name` means that the values is a terminal value, and it will be automatical ...
```
```suggestion

In general, structural tag templates support placeholders in the structural tags, which are in the form of `\{\{(var_name([])?)(\.var_name([])?)\}\}`, such as `{{A.name}}`, `{{A[].name}}`, `{{A.name[].other}}`. `var_name[]` means that the value is an array, and it will be automatically expanded with the users' input. `var_name` means that the values is a terminal value, and it will be automatically replaced with the input. Users can use `xgrammar.Grammar.from_structural_tag_template()` to generate a `xgrammar.Grammar`. Here, `template_json_str` is the structural tag templates, and kwargs are a series of values. For example, the structural tag template `stag` contains `{{A[].name}}` and `{{A[].age}}`, then the users can use `xgrammar.Grammar.from_structural_tag_template(stag, A=A)` to generate the grammar, where `A=[{"name":..., "age":...},.{"name":...,"age":...}, ...]`. This function will replace the placeholders automatically.
```

Seven-Streams committed:
add api.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the basic structure.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

feat:add a function to detect placeholder.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

feat:add the visit functions.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

update.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the basic design.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix the design.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish support for const string.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add fill functions for basic stags.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add to string functions.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add the detection for mingling.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix apis.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix mingle privacy.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish support for sequence.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish all the fillers except tagformat.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the template filler.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix api.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish python api.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix api.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix api.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the testing for const string.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix:fix jsonschemaformat and qwenformat.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add tests for jsonschemaformat.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish testing for qwen xml format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the check for regex.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the testing for basic types.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish basic tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add test for orformat.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix testing for sequence.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add the testing for mingled or and sequence.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix dummy placeholder name.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix dummy test.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish testing for mingled tagformat.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish basic testing for triggered tag template.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish all the combinatorial stags tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish part of builtin templates.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish qwen style template.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the templates.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix kimi format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add tests for deepseek and qwen_coder.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish testing for qwen.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the tests for harmony.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

support no tools provided.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

doc.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add doc.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

use picojson.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor the tostring.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor the regex.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor the replacement.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

ready to rewrite the expansion functions.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor expansion functions.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add a function to parse the placeholder.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add basic apis.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor the verification function.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor the verify function.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor the visit functions.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish all basic formats.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

feat:add the function to detect the expansion.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

feat:finish part of it.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish sequence.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Revert "finish."

This reverts commit 308ba13a2770544fbca7877b87257e3947426302.

Reapply "finish."

This reverts commit 0f360fa21293c903f7222a843073c9d46bcd8ea8.

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix api.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix layers.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix test.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix replace.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix the basic part.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix most of tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

refactor tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix original part.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the new apis.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add mixed test.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add mingled test.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish multi-layers.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add a nested test.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

doc.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix doc.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish the website refactor.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix tojson.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add from template method.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix multiple args in one const_string format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

add example codes in doc.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

finish testing for efficiency.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

fix test.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

remove the garbage file.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Seven-Streams committed:
update the tojson method.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Seven-Streams committed:
fix tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Seven-Streams committed:
fix tojson. add docs.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Seven-Streams committed:
finish tests.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Seven-Streams committed:
format.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>

Seven-Streams committed:
fix.

Signed-off-by: Yuchuan <yuchuan.7streams@gmail.com>
